### PR TITLE
bluetooth: hci: rpmsg: use ipc service library

### PR DESCRIPTION
**Pull request by kapi-no**
The Bluetooth HCI driver based on the RPMsg transport now uses the IPC
service module. The compatible Bluetooth sample - HCI RPMsg - has also
been migrated to the new IPC solution.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

### CODE REVIEW

**Review by jhedberg — DISMISSED** [review]

**Review by hubertmis — DISMISSED** [review]
Merging this PR should be synchronized with similar backend changes in https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/hal_nordic/nrf_802154/serialization/platform/nrf_802154_spinel_backend_ipc.c to make sure multiprotocol application are working correctly.

@e-rk, will you have a look at this?

**Review by alwa-nordic — DISMISSED** [review]

File: drivers/bluetooth/hci/Kconfig [hunk 1]
```diff
@@ -102,23 +112,6 @@ config BT_STM32_IPM_RX_STACK_SIZE
...
-	  are used in the HCI driver: channel 0 for TX and channel 1 for RX.
-
-if BT_RPMSG_NRF53
-
-choice RPMSG_SERVICE_MODE
-	default RPMSG_SERVICE_MODE_MASTER
-endchoice
-
-endif # BT_RPMSG_NRF53
-
```
Comment (Thalley) [inline drivers/bluetooth/hci/Kconfig:121]: Does this change mean that RPMSG is not automatically enabled for the nRF5340 CPUAPP?

**Comment by Thalley** [comment]
> The Bluetooth HCI driver based on the RPMsg transport now uses the IPC service module. The compatible Bluetooth sample - HCI RPMsg - has also been migrated to the new IPC solution.
> 

Is there any known sideeffect (positive or negative) of this change?

File: drivers/bluetooth/hci/Kconfig (hunk 1 above)
Comment (kapi-no) [reply drivers/bluetooth/hci/Kconfig:121]: No, it does not. I migrated the HCI driver to a transport module (IPC service) with a better abstraction. It is no longer necessary to have SoC specific code.

File: drivers/bluetooth/hci/rpmsg.c [hunk 2]
```diff
@@ -257,13 +259,45 @@ static int bt_rpmsg_send(struct net_buf *buf)
...
+
 	BT_DBG("");
 
-	while (!bt_rpmsg_platform_endpoint_is_bound()) {
-		k_sleep(K_MSEC(1));
+	err = ipc_service_open_instance(hci_ipc_instance);
+	if (err) {
+		BT_ERR("IPC service instance initialization failed: %d\n", err);
+		return err;
 	}
```
[the remark this answers is not in the document]
Comment (kapi-no) [reply drivers/bluetooth/hci/rpmsg.c:291]: Yes, thanks for the tip. I will correct it.

File: samples/bluetooth/hci_rpmsg/src/main.c [hunk 3]
```diff
@@ -266,6 +272,19 @@ void main(void)
...
+	if (err) {
+		LOG_ERR("IPC service instance initialization failed: %d\n", err);
+	}
+
+	err = ipc_service_register_endpoint(hci_ipc_instance, &hci_ept, &hci_ept_cfg);
+	if (err) {
+		LOG_ERR("Registering endpoint failed with %d", err);
+	}
+
+	k_sem_take(&ipc_bound_sem, K_FOREVER);
```
[the remark this answers is not in the document]
Comment (kapi-no) [reply samples/bluetooth/hci_rpmsg/src/main.c:286]: yes, it makes sense to add such security measures on both ends (HCI driver and sample)

**Comment by kapi-no** [comment]
> > The Bluetooth HCI driver based on the RPMsg transport now uses the IPC service module. The compatible Bluetooth sample - HCI RPMsg - has also been migrated to the new IPC solution.
> 
> Is there any known sideeffect (positive or negative) of this change?

@Thalley, yes, there is.

This module has multi-instance support. You can assign different regions of shared memory to each IPC instance via DTS configuration.

**Comment by kapi-no** [comment]
Addressed outstanding review comments. Passing CI is still a work in progress.

**Comment by kapi-no** [comment]
This PR depends on the following PRs:
https://github.com/zephyrproject-rtos/zephyr/pull/43370
https://github.com/zephyrproject-rtos/zephyr/pull/43325

**Review by e-rk — APPROVED** [review]

**Comment by carlescufi** [comment]
> This PR depends on the following PRs: #43370 #43325

Both are merged now

**Comment by carlescufi** [comment]
> > The Bluetooth HCI driver based on the RPMsg transport now uses the IPC service module. The compatible Bluetooth sample - HCI RPMsg - has also been migrated to the new IPC solution.
> 
> Is there any known sideeffect (positive or negative) of this change?

Aside from what @kapi-no points out, there is no longer need for an nRF53-specific file under `drivers/bluetooth/hci` now, it's all generic!

**Review by carlescufi — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**